File: app/src/main/java/com/buzzybrains/mvvmarchitecture/data/NoteRepository.java
package com.buzzybrains.mvvmarchitecture.data;

import android.app.Application;

import com.buzzybrains.mvvmarchitecture.asynctask.DeleteAllNoteAsyncTask;
import com.buzzybrains.mvvmarchitecture.asynctask.DeleteNoteAsyncTask;
import com.buzzybrains.mvvmarchitecture.asynctask.InsertNoteAsyncTask;
import com.buzzybrains.mvvmarchitecture.asynctask.UpdateNoteAsyncTask;
import com.buzzybrains.mvvmarchitecture.model.Note;
import com.buzzybrains.mvvmarchitecture.serviceimpl.RemoteSyncServiceImpl;
import com.buzzybrains.mvvmarchitecture.services.INoteService;
import com.buzzybrains.mvvmarchitecture.services.RemoteSyncService;

import java.util.List;

import androidx.lifecycle.LiveData;

public class NoteRepository implements INoteService {

    private NoteDao noteDao;

    private LiveData<List<Note>> allNotes;

    private RemoteSyncService remoteSyncService;


    public NoteRepository(Application application) {
        NoteDatabase noteDatabase = NoteDatabase.getInstance(application);

        noteDao = noteDatabase.noteDao();

        allNotes = noteDao.getAllNotes();


        remoteSyncService = new RemoteSyncServiceImpl(this);

    }

    public void syncNoteData(Note note) {
        remoteSyncService.syncNote(note);
    }


    public void insert(Note note) {
        new InsertNoteAsyncTask(noteDao, this).execute(note);
    }

    public void update(Note note) {


        new UpdateNoteAsyncTask(noteDao, this).execute(note);
    }


    public void delete(Note note) {
        new DeleteNoteAsyncTask(noteDao).execute(note);
    }

    public void deleteAllNotes() {
        new DeleteAllNoteAsyncTask(noteDao).execute();
    }

    public LiveData<List<Note>> getAllNotes() {
        return allNotes;
    }


    public List<Note> getAsyncNotes() {
        return noteDao.getAsyncNotes();
    }


    @Override
    public void getSyncSuccessNoteObject(Note note) {
        note.setSync(true);

        new UpdateNoteAsyncTask(noteDao, this).execute(note);

    }

    @Override
    public void getSyncFailureNoteObject(Note note) {

    }


}

File: app/src/main/java/com/buzzybrains/mvvmarchitecture/data/NoteRepository.kt
package com.buzzybrains.mvvmarchitecture.data

import android.app.Application

import com.buzzybrains.mvvmarchitecture.asynctask.DeleteAllNoteAsyncTask
import com.buzzybrains.mvvmarchitecture.asynctask.DeleteNoteAsyncTask
import com.buzzybrains.mvvmarchitecture.asynctask.InsertNoteAsyncTask
import com.buzzybrains.mvvmarchitecture.asynctask.UpdateNoteAsyncTask
import com.buzzybrains.mvvmarchitecture.model.Note
import com.buzzybrains.mvvmarchitecture.serviceimpl.SyncServiceImpl
import com.buzzybrains.mvvmarchitecture.services.INoteService
import com.buzzybrains.mvvmarchitecture.services.ISyncService

import androidx.lifecycle.LiveData

class NoteRepository(application: Application) : INoteService {

    private val noteDao: NoteDao

    val allNotes: LiveData<List<Note>>

    private val remoteSyncService: ISyncService


    val asyncNotes: List<Note>
        get() = noteDao.asyncNotes


    init {
        val noteDatabase = NoteDatabase.getInstance(application)

        noteDao = noteDatabase.noteDao()

        allNotes = noteDao.allNotes

        remoteSyncService = SyncServiceImpl(this)

    }

    fun syncNoteData(note: Note) {
        remoteSyncService.syncNote(note)
    }


    fun insert(note: Note) {
        InsertNoteAsyncTask(noteDao, this).execute(note)
    }

    fun update(note: Note) {


        UpdateNoteAsyncTask(noteDao, this).execute(note)
    }


    fun delete(note: Note) {
        DeleteNoteAsyncTask(noteDao).execute(note)
    }

    fun deleteAllNotes() {
        DeleteAllNoteAsyncTask(noteDao).execute()
    }


    override fun getSyncSuccessNoteObject(note: Note) {
        note.isSync = true

        UpdateNoteAsyncTask(noteDao, this).execute(note)

    }

    override fun getSyncFailureNoteObject(note: Note) {

    }


}
